Tests/AsyncHTTPClientTests/TransactionTests.swift

```diff
@@ -541,16 +541,26 @@ final class TransactionTests: XCTestCase {
 // tasks. Since we want to wait for things to happen in tests, we need to `async let`, which creates
 // implicit tasks. Therefore we need to wrap our iterator struct.
 @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
-actor SharedIterator<Iterator: AsyncIteratorProtocol> {
-    private var iterator: Iterator
+actor SharedIterator<Wrapped: AsyncSequence> where Wrapped.Element: Sendable {
+    private var wrappedIterator: Wrapped.AsyncIterator
+    private var nextCallInProgress: Bool = false
 
-    init(_ iterator: Iterator) {
-        self.iterator = iterator
+    init(_ sequence: Wrapped) {
+        self.wrappedIterator = sequence.makeAsyncIterator()
     }
 
-    func next() async throws -> Iterator.Element? {
-        var iter = self.iterator
-        defer { self.iterator = iter }
+    func next() async throws -> Wrapped.Element? {
+        precondition(self.nextCallInProgress == false)
+        self.nextCallInProgress = true
```

Comment:
This construct is quite unsafe and doesn't support reentrancy. `nextCallInProgress` makes it a bit safe and will crash if next is called while another call to next is still in progress.